File: ios/MendixNative/AppPreferences/AppPreferences.m
#import "AppPreferences.h"
#import "WarningsFilter.h"
#import "AppUrl.h"

@implementation AppPreferences
NSString *appUrlKey = @"ApplicationUrl";
NSString *devModeEnabledKey = @"DevModeEnabled";
NSString *clearDataEnabledKey = @"ClearData";
NSString *remoteDebuggingEnabledKey = @"RemoteDebuggingEnabled";
NSString *remoteDebuggingPackagerPortKey = @"RemoteDebuggingPackagerPort";
NSString *elementInspectorDebugKey = @"showInspector";

+ (NSString*) getAppUrl {
    return [[NSUserDefaults standardUserDefaults] stringForKey:appUrlKey];
}

+ (void) setAppUrl:(NSString*) url {
    [[NSUserDefaults standardUserDefaults] setObject:url forKey:appUrlKey];
}

+ (BOOL) devModeEnabled {
    return [[NSUserDefaults standardUserDefaults] boolForKey:devModeEnabledKey];
}

+ (void) devMode:(BOOL)enable {
    [[NSUserDefaults standardUserDefaults] setBool:enable forKey:devModeEnabledKey];
}

+ (BOOL) remoteDebuggingEnabled {
    return [[NSUserDefaults standardUserDefaults] boolForKey:remoteDebuggingEnabledKey];
}

+ (void) remoteDebugging:(BOOL)enable {
    [[NSUserDefaults standardUserDefaults] setBool:enable forKey:remoteDebuggingEnabledKey];
}

+ (int) getRemoteDebuggingPackagerPort {
    int port = (int) [[NSUserDefaults standardUserDefaults] integerForKey:remoteDebuggingPackagerPortKey];
    return port != 0 ? port : [AppUrl defaultPackagerPort];
}

+ (void) setRemoteDebuggingPackagerPort: (NSInteger)port {
    [[NSUserDefaults standardUserDefaults] setInteger:port forKey:remoteDebuggingPackagerPortKey];
}

+ (BOOL) isElementInspectorEnabled {
    return [[NSUserDefaults standardUserDefaults] boolForKey:elementInspectorDebugKey];
}

+ (void) setElementInspector:(BOOL)enable {
    [[NSUserDefaults standardUserDefaults] setBool:enable forKey:elementInspectorDebugKey];
}

@end

File: ios/Modules/AppPreferences/AppPreferences.swift
import Foundation

public class AppPreferences: NSObject {
    @UserDefault(key: "ApplicationUrl", defaultValue: nil)
    static var _appUrl: String?
    
    @UserDefault(key: "DevModeEnabled", defaultValue: false)
    static var _devModeEnabled: Bool
    
    @UserDefault(key: "RemoteDebuggingEnabled", defaultValue: false)
    static var _remoteDebuggingEnabled: Bool
    
    @UserDefault(key: "showInspector", defaultValue: false)
    static var _elementInspectorEnabled: Bool
    
    @UserDefault(key: "RemoteDebuggingPackagerPort", defaultValue: AppUrl.defaultPackagerPort)
    private static var _packagerPort: Int
    
    public static var remoteDebuggingPackagerPort: Int {
        get { AppUrl.ensurePort(_packagerPort) }
        set { _packagerPort = newValue }
    }
    
    public static var appUrl: String? {
        get { _appUrl }
        set { _appUrl = newValue }
    }
    
    public static var devModeEnabled: Bool {
        get { _devModeEnabled }
        set { _devModeEnabled = newValue }
    }
    
    public static var remoteDebuggingEnabled: Bool {
        get { _remoteDebuggingEnabled }
        set { _remoteDebuggingEnabled = newValue }
    }
    
    public static var elementInspectorEnabled: Bool {
        get { _elementInspectorEnabled }
        set { _elementInspectorEnabled = newValue }
    }
    
    public static var safeAppUrl: String {
        return appUrl ?? ""
    }
    
}
